### PR TITLE
Add graticule labels feature

### LINKED ISSUE #3210: labels for ol.Graticule
Hey, 

I migrated from OL2 to OL3 on my app and realize that ol.Graticule have not more support for labels!?
How can I fix it asap?

Thanks a lot! 
Gaston

PS: I love OL! is the best map tool I ever  use

### LINKED ISSUE #4307: Graticule Labelled
Hi There,

I am trying to use the graticule control in my project using Open Layers 3.10.1, however since upgrading from Open Layers 2, I cannot seem to get graticule to display labels. There use to be a "labelled" flag in the control which would allow for this to work.

Is there any chance that this could be added to the graticule control?

Cheers
Rod

### PR DESCRIPTION
**Pull request by Brictarus**
Hello, this is a pull request with an implementation of graticule labels. Please review ! :). 
Fixes #3210
Fixes #4307 (issues are bound)

### CODE REVIEW

**Comment by Brictarus** [comment]
See example below for preview : 
2015-11-03_16h39_38

File: externs/olx.js [hunk 1]
```diff
@@ -101,6 +101,11 @@ olx.LogoOptions.prototype.src;
  * @typedef {{map: (ol.Map|undefined),
  *     maxLines: (number|undefined),
  *     strokeStyle: (ol.style.Stroke|undefined),
+ *     showLabels: (boolean|undefined),
+ *     lonLabelFormatter: (Function|undefined),
+ *     latLabelFormatter: (Function|undefined),
+ *     lonLabelPosition: (number|undefined),
+ *     latLabelPosition: (number|undefined),
```
Comment (petrsloup) [inline externs/olx.js]: You should add documentation for these options below.
Also the formatter functions can be more strongly typed (`function(number):string`), right?

File: src/ol/graticule.js [hunk 2]
```diff
@@ -134,6 +161,41 @@ ol.Graticule = function(opt_options) {
...
+
+  /**
+   * @type {number} between 0.0 and 1.0
+   * @private
+   */
+  this.lonLabelPosition_ = options.lonLabelPosition !== undefined ?
+      ol.math.clamp(options.lonLabelPosition, 0.0, 1.0) : 1.0;
+
+  /**
+   * @type {?Function}
```
Comment (petrsloup) [inline src/ol/graticule.js]: Again, the function type can be stronger here as well (the same as above)

File: src/ol/graticule.js [hunk 3]
```diff
@@ -134,6 +161,41 @@ ol.Graticule = function(opt_options) {
...
+   */
+  this.latLabelFormatter_ = options.latLabelFormatter !== undefined ?
+      options.latLabelFormatter : null;
+
+  /**
+   * @type {number} between 0.0 and 1.0
+   * @private
+   */
+  this.latLabelPosition_ = options.latLabelPosition !== undefined ?
+      ol.math.clamp(options.latLabelPosition, 0.0, 1.0) : 1.0;
```
Comment (petrsloup) [inline src/ol/graticule.js]: Does it actually make sense to have default label positions `1.0` ?
This value makes the labels float outside(/on the edge of) the viewport, so anyone who wants to use this would have to specify more sensible value anyway..

**Comment by marcjansen** [comment]
For the formatter functions it may make sense to use `ol.coordinate.format`.

**Comment by petrsloup** [comment]
@marcjansen Yes, that would also be possible, but the behavior here is slightly different, since the latitude and longitude values are formatted independently (not as a coordinate).

@Brictarus Thanks for this PR ! Would it be possible to improve the behavior of the labels so they do not float away from the graticule if the map is zoomed out?

**Comment by Brictarus** [comment]
Hello guys,

@petrsloup i don't understand what you mean by
> improve the behavior of the labels so they do not float away from the graticule if the map is zoomed out

Can you summarize the minimum required improvements to make this PR accepted ? (I don't have much time to offer thses days...) ?

I just discover that the examples works only with mode=raw. Do you know why ? how can i fix that ?

Thanks.

**Comment by petrsloup** [comment]
@Brictarus This is the state I was trying to describe:
screenshot from 2015-11-09 14-56-29
I think this could be improved (the labels can be closer to the graticule).

**Comment by eblondel** [comment]
@Brictarus I've provided a pull request to your fork that fixes what has been highlighted by @petrsloup , adding more options to control the graticule labels. I would be grateful if you could merge it to your initial pull request so others can assess it here, and hopefully make it part of ol3. Thanks!

**Comment by cityant** [comment]
When will the feature been picked into next release? It's great and looking forward to getting it soon.

**Comment by ahocevar** [comment]
@Brictarus, can you please merge @eblondel's pull request into your branch so we can continue the review here?

**Comment by ahocevar** [comment]
Fixed with #6737.
